schema/spacewalk/upgrade/susemanager-schema-4.4.8-to-susemanager-schema-4.4.9/203-add-changelogdata-index.sql

```diff
@@ -1,27 +1,28 @@
 create or replace function remove_duplicate_changelogdata()
 returns void as
 $$
-declare original record;
-declare duplicate record;
 begin
-    for original in select min(id) as id
+    update rhnpackagechangelogrec
+        set changelog_data_id = duplicates.keep_id
+        from (
+            select
+                min(id) as keep_id,
+                array_agg(id) as duplicate_ids
             from rhnpackagechangelogdata
             group by name, text, time
-            having count(*) > 1 loop
-        for duplicate in select data2.id
-                from rhnpackagechangelogdata data1, rhnpackagechangelogdata data2
-                where data1.name = data2.name
-                and data1.text = data2.text
-                and data1.time = data2.time
-                and data1.id != data2.id
-                and data1.id = original.id loop
-            update rhnpackagechangelogrec
-                set changelog_data_id = original.id
-                where changelog_data_id = duplicate.id;
-            delete from rhnpackagechangelogdata
-                where id = duplicate.id;
-        end loop;
-    end loop;
+            having count(*) > 1
+        ) as duplicates
+        where changelog_data_id = any(duplicates.duplicate_ids);
+
+        delete from rhnpackagechangelogdata
+        where id in (select unnest(duplicate_ids) from (
+            select
+                min(id) as keep_id,
+                array_agg(id) as duplicate_ids
+            from rhnpackagechangelogdata
+            group by name, text, time
+            having count(*) > 1
+        ) as d where d.keep_id <> any(d.duplicate_ids));
```

Comment:
Are we sure that this is not removing the keep_id as well?
This `<> ANY` is not clear to me what it exactly does.

Reply:
You're right. That's bogus. Fixing.